File: fe/src/services/stock.ts
import {requestHelpers} from "../helpers";
import services from "./request";
import {IDataResponse, StockInfo, StockPrice} from "../interfaces";
import {StockSymbol} from "../interfaces";


function searchSymbol(keyword: string, page?: number, pageSize?: number): Promise<IDataResponse<StockSymbol[]>>  {
  return services.get(`/stock/search`, {
    params: {
      keyword,
      page: page ?? 1,
      page_size: pageSize ?? 20,
    }
  }).then(requestHelpers.getResult<StockSymbol[]>);
}

const getStockHistory = async (symbol: string, start?: string, end?: string, page?: number, pageSize?: number) => {
  return services.get(`/stock/history`, {
    params: {
      symbol,
      start,
      end,
      page: page ?? 1,
      page_size: pageSize ?? 5,
    }
  }).then(requestHelpers.getResult<StockPrice[]>);
}

const getStockInfo = async (symbol: string) => {
  return services.get(`/stock/info/${symbol}`).then(requestHelpers.getResult<StockInfo>);
}

const addToInterestList = async (symbol: string) => {
  return services.post(`/stock/interest/${symbol}`).then(requestHelpers.getResult);
}

const deleteFromInterestList = async (symbol: string) => {
  return services.delete(`/stock/interest/${symbol}`).then(requestHelpers.getResult);
}

const getInterestList = async () => {
  return services.get(`/stock/interest`).then(requestHelpers.getResult<StockInfo[]>);
}

const stockServices = {
  searchSymbol,
  getStockHistory,
  addToInterestList,
  deleteFromInterestList,
  getInterestList,
  getStockInfo,
}

export default stockServices;

File: be/app/apis/endpoints/stock.py
from fastapi import APIRouter, Depends

from app.core.security import get_user_info
from app.schemas import PagingDataResponse, Stock, StockHistoryParam, Symbol, SearchSymbolParam, ResponseBase, \
    StockInfo, StockDetailInfo, DataResponse
from app.services.stock import StockService

router = APIRouter()


@router.get("/search", response_model=PagingDataResponse[Symbol])
def search_stock(params=Depends(SearchSymbolParam)):
    result, metadata = StockService.search_symbol(params)
    return PagingDataResponse().success_response(result, metadata)


@router.get("/history", response_model=PagingDataResponse[Stock])
def get_stock_history(history_request=Depends(StockHistoryParam)):
    result, metadata = StockService.get_history(history_request)
    return PagingDataResponse().success_response(result, metadata)


@router.get("/info/{symbol}", response_model=DataResponse[StockDetailInfo])
def get_stock_info(symbol: str):
    result = StockService.get_stock_info(symbol)
    return DataResponse().success_response(result)


@router.get("/interest", response_model=PagingDataResponse[StockInfo])
async def get_interest_stock(user_id=Depends(get_user_info)):
    result, metadata = await StockService.get_interest_stock(user_id)
    return PagingDataResponse().success_response(result, metadata)


@router.post("/interest/{symbol}", response_model=ResponseBase)
def add_interest_stock(symbol: str, user_id=Depends(get_user_info)):
    StockService.add_interest_stock(user_id, symbol)
    return ResponseBase().success_response()


@router.delete("/interest/{symbol}", response_model=ResponseBase)
def remove_interest_stock(symbol: str, user_id=Depends(get_user_info)):
    StockService.remove_interest_stock(user_id, symbol)
    return ResponseBase().success_response()
